cache/cache.go

```diff
@@ -0,0 +1,66 @@
+package cache
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/rogpeppe/go-internal/lockedfile"
+	"golang.org/x/exp/slog"
+)
+
+func Add(key string, addItem func() error) error {
+	if err := os.MkdirAll(key, 0700); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to create directory %s", key))
+	}
+
+	lockFpath := filepath.Join(key, ".lock")
+	slog.Debug("taking lock", "key", lockFpath)
+	lock, err := lockedfile.Create(lockFpath)
+	slog.Debug("took lock", "key", lockFpath)
+
+	if err != nil {
+		return errors.Wrap(err, "failed to take file lock")
+	}
+	defer func() {
+		if err := lock.Close(); err != nil {
+			slog.Error("failed to release lock", "key", lockFpath, "error", err)
+		}
+		slog.Debug("released lock", "key", lockFpath)
+	}()
+	// If data is already present, return
+	if _, err := os.Stat(filepath.Join(key, ".success")); err == nil {
+		return nil
```

Comment:
Nit: I think it is easy to reason what's going on if you used `.started` and `.done` as marker files for concurrency handling.

Reply:
.started is not required. I have changed completion marker file to .done

Reply:
You are using `.lock` for the `.started` marker. That's what I meant to rename in my prev comment.